test/integration/nighthawk_test_server.py

```diff
@@ -30,41 +33,82 @@ def __init__(self, server_binary_path, config_template_path, server_ip, ip_versi
     self.server_process = None
     self.server_ip = server_ip
     self.socket_type = socket.AF_INET6 if ip_version == IpVersion.IPV6 else socket.AF_INET
-    self.server_port = -1
-    self.admin_port = -1
     self.admin_address_path = ""
     self.parameterized_config_path = ""
     self.instance_id = str(random.randint(1, 1024 * 1024 * 1024))
     self.parameters = parameters
     self.server_binary_config_path_arg = server_binary_config_path_arg
-
     self.parameters["server_ip"] = self.server_ip
-    with open(self.config_template_path) as f:
-      config = Template(f.read())
-      config = config.substitute(self.parameters)
-      logging.info("Parameterized server configuration: %s", config)
-
-    with tempfile.NamedTemporaryFile(mode="w", delete=False, suffix=".yaml") as tmp:
+    self.docker_image = os.getenv("NH_DOCKER_IMAGE", "")
+    self.tmpdir = os.path.join(os.getenv("TMPDIR", "/tmp/nighthawk_benchmark/"), tag + "/")
+    self.parameters["tmpdir"] = self.tmpdir
+    self.parameters["tag"] = tag
+
+    def substitute_yaml_values(obj, params):
+      if isinstance(obj, dict):
+        for k, v in obj.items():
+          obj[k] = substitute_yaml_values(v, params)
+      elif isinstance(obj, list):
+        for i in range(len(obj)):
+          obj[i] = substitute_yaml_values(obj[i], params)
+      else:
+        if isinstance(obj, str):
+          # Inspect string values and substitute where applicable.
+          INJECT_RUNFILE_MARKER = '@inject-runfile:'
+          if obj[0] == '$':
+            return Template(obj).substitute(params)
+          elif obj.startswith(INJECT_RUNFILE_MARKER):
+            r = runfiles.Create()
+            with open(r.Rlocation(obj[len(INJECT_RUNFILE_MARKER):].strip()), 'r') as file:
+              return file.read()
+      return obj
+
+    r = runfiles.Create()
+    with open(r.Rlocation(self.config_template_path)) as f:
```

Comment:
Why not just make the entire config a giant `string.Template` or Python f-string? That way you don't need to do YAML traversal.

Reply:
Done